### PR TITLE
Resolve simultaneous deployments with rebase

### LINKED ISSUE #1052: Force pushing risks losing data
It looks like this action implements its changes by force-pushing to the target branch:

https://github.com/JamesIves/github-pages-deploy-action/blob/226f2c44b58c32b6f07f5835346633e66fc069db/src/git.ts#L220-L221

I believe this works for the vast majority of use cases, but not all of them. In the case where two deployments to different locations on the same branch are initiated at the same time (or perhaps just very close to each other), changes could be destroyed.

<details>
<summary><b>Details about my specific use case</b></summary>

Consider a use case where this action is being used to create multiple deployments in different directories in a single branch - for example, a main `gh-pages` deployment and a preview deployment for a pull request:

```
branch: gh-pages
/
├─ pr-preview/
│  └─ pr-8/
│     └─ <preview deployment>
└─ <main deployment>
```

This repository would have one workflow that creates the main deployment on push to the main branch:

```yml
# .github/workflows/deploy.yml
name: build
on:
  push:
    branches: [main]
jobs:
  deploy:
    steps:
      - uses: actions/checkout@v2
      - run: npm install && npm run build
      - uses: JamesIves/github-pages-deploy-action@v4
        with:
          branch: gh-pages
          folder: dist
          clean-exclude: pr-preview/
```

And another that handles preview deployments for pull requests - both setting them up for `opened`/`closed`/`synchronize` events, and removing them on `closed` events:

```yml
# .github/workflows/preview.yml
on:
  pull_request:
    types:
      - opened
      - reopened
      - synchronize
      - closed
jobs:
  deploy-preview:
    if: github.event.pull_request.action != 'closed'
    steps:
      - uses: actions/checkout@v2
      - run: npm install && npm run build
      - uses: JamesIves/github-pages-deploy-action@v4
        with:
          branch: gh-pages
          folder: dist
          target-folder: pr-preview/pr-${{ github.event.pull_request.number }}
  remove-preview:
    if: github.event.pull_request.action == 'closed'
    steps:
      - run: echo "empty_dir=$(mktemp -d)" >> $GITHUB_ENV
      - uses: JamesIves/github-pages-deploy-action@v4
        with:
          branch: gh-pages
          folder: ${{ env.empty_dir }}
          target-folder: pr-preview/pr-${{ github.event.pull_request.number }}
```

This is where the action's current assumption that all deployments will neatly overwrite the previous deployment breaks down. When a pull request is merged to the main branch, the pull request's `closed` event and the main branch's `push` event will fire at the same time. This causes two simultaneous deployments.

Both deployments will pull and secure a reference to the current SHA of the `gh-pages` branch. They'll prepare their commit, and then push. Normally, one would be successful and the other would fail, but in this case both force-push their changes and one of the commits will be destroyed.

In my case, I merged a pull request, the preview for the pull request was successfully removed, and the main deployment was lost.

</details>

I believe this could be resolved by changing how the action pushes its changes. Rather than force-push the deployment, I think it should attempt a regular push then, if that fails, pull the remote and push again just like a human would.

An implementation that does this forever until it pushes successfully could look something like this:

```js
let rejected
do {
  let pushResult = await execute(
    `git push --porcelain ${action.repositoryPath} ${temporaryDeploymentBranch}:${action.branch}`
  )
  rejected = pushResult.includes(`[rejected]`)
  if (rejected) await execute(
    `git pull --rebase --no-edit ${action.repositoryPath} ${action.branch}:${temporaryDeploymentBranch}`
  )
} while (rejected)
```

Do you think this would work / is a good idea? Have I misdiagnosed the issue; or is there a better way of resolving it?

I'm happy to work on this as a PR - I need this functionality, so I'll be mucking about in a fork regardless. Please do let me know if you have any thoughts.

### PR DESCRIPTION
**Pull request by rossjrw**
Fixes #1052.

Deployed version of fork: https://github.com/rossjrw/github-pages-deploy-action/tree/dev-merge-test

Use with `uses: rossjrw/github-pages-deploy-action@dev-merge-test`

This PR adds a new parameter, `force`, which defaults to true. When set to true the action will deploy all changes with `git push --force`. When set to any other value, the action will do the following:

1. Attempt to deploy changes with `git push`
2. If changes are rejected, `git fetch` the remote branch and `git rebase` local changes onto it.
3. Repeat up to a set number of times, currently 3.

Test repository: https://github.com/rossjrw/temp-pages-deploy-action-merge-test

I've invited you as a collaborator on both repositories in case there are any changes you'd like to make to any branch beyond the single one that 'Allow edits by maintainers' grants.

I have not yet updated tests (edit: I have updated existing tests, which now pass, but have not added any additional ones) or documentation. I've also not tested to see what happens when used in combination with e.g. `singleCommit`.

### CODE REVIEW

**Comment by JamesIves** [comment]
Im planning to review this over the weekend. Thanks for the contribution!

File: action.yml [hunk 1]
```diff
@@ -60,6 +60,11 @@ inputs:
     description: 'Do not actually push back, but use `--dry-run` on `git push` invocations insead.'
     required: false
 
+  force:
```
Comment (JamesIves) [inline action.yml:63]: Can we add this to the README under the `Optional Choices` section?

**Review by JamesIves — COMMENTED** [review]
Looks great I couldn't find any bugs after my first look. Do you mind adding a test case for this into the `workflows/integration.yml` file? We should add a couple of cases that validate some of the different workflow permutations such as `single-commit` and `repository-name`.  For the repository name you can point it to `MontezumaIves/lab`

**Comment by rossjrw** [comment]
Just noting that I intend to get around to implementing integration tests later this week.

**Comment by rossjrw** [comment]
@JamesIves I've created an integration test, though I'm not sure what your preferences are regarding what it should actually do, and I'm not sure that I'm able to even run the test myself. I invite you to look over the jobs I added to the workflow and adjust as appropriate. I suspect you may need to take the reins on the permutations you mentioned.

**Comment by JamesIves** [comment]
Sounds good, I'll take it from here. I'm traveling at the minute but will get to this once I'm home.

**Pull request merged**